### ISSUE TITLE
Origin header is not set to WebSocket when SIP.js is used from node server

### ISSUE THREAD

**Issue by vkuzmenok** [onsip/SIP.js#535]
Could you add a possibility to specify Origin header for WebSocket  creation, when SIP.js is used from node server? Out of the box, ws module does not set Origin header and there is no way to overcome this problem, unless you specify options when you call new WebSocket constructor.

In our setup we were able to resolve this by fixing the following line in the sip.js

`this.ws = new WebSocket(this.server.ws_uri, 'sip', {origin: 'foo'});`

Obviously, this is not the right way to do this, so it would be nice to have a possibility to provide web socket options as a configuration part of the SIP.UA constructor

**Comment by james-criscuolo** [comment]
Hi @vkuzmenok,
  We are currently working on a way to make the transport layer more generic and editable, similar to what has occured with the session description handler. I am not sure if that will be in the upcoming release, but it should be in soon after, and at that point this should certainly be possible.

Thanks,
James

**Comment by vkuzmenok** [comment]
Thanks James. 

Do you have a timeline when you expect next releases to come out?

Vladimir

**Comment by james-criscuolo** [comment]
The next release will be in the next week or two, if all goes as planned. I am not certain on the following release, but once the work in question is done it would release.

**Comment by vkuzmenok** [comment]
Does this mean, you will provide an option to specify the transport in the UA contructor?

**Comment by egreenmachine** [comment]
That is the plan. Should be extremely similar to how you can specify your own Session Description Handler.

**Comment by vkuzmenok** [comment]
Sounds great! Thanks!

**Comment by JimGreenberg** [comment]
An update on this: the 0.11.0 release represents a huge refactor for the transport code- with the purpose of making this type of thing a lot easier. The UA now has a configuration parameter transportConstructor where you can give it a user-defined Transport (as @egreenmachine said). For this case it should be 99.9% a copy/paste job from the default websocket transport, where you'll be adding an extra argument to the `new Websocket` call and an additional configuration parameter on the new `Transport` you define. This might seem like overkill- to create a whole new file just for this one simple change, but it allows us to not have to support the nonstandard API while also giving you the flexibility to have the functionality.

As always, if you need help we have a [mailing](https://groups.google.com/forum/#!forum/sip_js) list that gets checked frequently.
